pkg/prom/wal/series.go

```diff
@@ -200,40 +194,25 @@ func (s *stripeSeries) getByID(id uint64) *memSeries {
 	return series
 }
 
-// getByHash looks up a series that was saved via saveLabels.
 func (s *stripeSeries) getByHash(hash uint64, lset labels.Labels) *memSeries {
 	i := hash & uint64(s.size-1)
 
 	s.locks[i].RLock()
-	defer s.locks[i].RUnlock()
-
-	ref, ok := s.hashes[i].get(hash, lset)
-	if !ok {
-		return nil
-	}
-
-	j := ref & uint64(s.size-1)
-	if j != i {
-		s.locks[j].RLock()
-		defer s.locks[j].RUnlock()
-	}
+	series := s.hashes[i].get(hash, lset)
+	s.locks[i].RUnlock()
 
-	return s.series[j][ref]
```

Comment:
This is the bulk of the bugfix. This removes the need for obtaining a second lock by changing hashes to store a memSeries directly. All the other code is centered around dealing with how memseries now holds the labels rather than the hash.